### PR TITLE
[SQOOP-3224] Binary transfer mode

### PR DESCRIPTION
**Pull request by christeoh**

### CODE REVIEW

File: src/java/org/apache/sqoop/mapreduce/mainframe/MainframeConfiguration.java [hunk 1]
```diff
@@ -33,4 +33,13 @@
...
   public static final String MAINFRAME_FTP_FILE_ENTRY_PARSER_CLASSNAME = "org.apache.sqoop.mapreduce.mainframe.MainframeFTPFileEntryParser";
+
+  public static final String MAINFRAME_FTP_TRANSFER_MODE = "mainframe.ftp.transfermode";
+
+  public static final String MAINFRAME_FTP_TRANSFER_MODE_ASCII = "ascii";
+
+  public static final String MAINFRAME_FTP_TRANSFER_MODE_BINARY = "binary";
+
+  // this is the buffer size used when doing binary ftp transfers from mainframe
+  public static final Integer MAINFRAME_FTP_TRANSFER_BINARY_BUFFER = 32760;
```
Comment (fszabo2) [inline src/java/org/apache/sqoop/mapreduce/mainframe/MainframeConfiguration.java]: Is this value always a good choice? I would make it configurable (if it isn't already), and use this value as default if said configuration is not present, so the user has a choice here.

File: src/java/org/apache/sqoop/util/MainframeFTPClientUtils.java [hunk 2]
```diff
@@ -207,8 +208,18 @@ public static FTPClient getFTPConnection(Configuration conf)
         throw new IOException("Could not login to server " + server
             + ":" + ftp.getReplyString());
       }
-      // set ASCII transfer mode
-      ftp.setFileType(FTP.ASCII_FILE_TYPE);
+      // set transfer mode
+      String transferMode = conf.get(MainframeConfiguration.MAINFRAME_FTP_TRANSFER_MODE);
```
Comment (fszabo2) [inline src/java/org/apache/sqoop/util/MainframeFTPClientUtils.java:212]: The whole getFTPConnection method is already very long and does a bunch of steps. You could consider refactoring it into smaller methods for readability.

File: src/test/org/apache/sqoop/util/TestMainframeFTPClientUtils.java [hunk 3]
```diff
@@ -362,4 +363,29 @@ public void testPartitionedDatasetsShouldReturnAllFiles() {
       Assert.fail(ioeString);
     }
   }
+  @Test
+  public void testBinaryTransferMode() throws IOException {
+    final String EXPECTED_RESPONSE = "200 Representation type is Image";
+    final int EXPECTED_RESPONSE_CODE = 200;
+    when(mockFTPClient.login("user", "pssword")).thenReturn(true);
```
Comment (fszabo2) [inline src/test/org/apache/sqoop/util/TestMainframeFTPClientUtils.java:370]: is the typo "pssword" intentional? Though won't cause any trouble I guess...

**Review by fszabo2 — COMMENTED** [review]
Hi Chris,

I saw that this review started on ReviewBoard originally, perhaps it would be a good idea to finish it there, so we'll see the history in one place.

Anyway, I've added a few comments below before realizing this :)

File: src/java/org/apache/sqoop/util/MainframeFTPClientUtils.java (hunk 2 above)
Comment (christeoh) [reply src/java/org/apache/sqoop/util/MainframeFTPClientUtils.java:212]: I'm not sure how you would refactor it. I can give it a try and see if it is what you're looking for.

**Pull request closed without merging**